### DISCUSSION TITLE
Iterate over list of string using split or strings.Split

### DISCUSSION THREAD

**Discussion by ghost** [hairyhenderson/gomplate#1535] [Q&A] [answered]
Hello,

We are trying to iterate over the list of strings to generate required AWS providers for Terraform.

The most appropriate way to do it probably by using [strings.Split](https://docs.gomplate.ca/functions/strings/#strings-split) or [split](https://docs.gomplate.ca/functions/strings/#split) functions with a `range`. And it looks like `strings.Split` works only with a pipe-line, but in documentation we see both cases:
<img width="946" alt="Screen Shot 2022-10-22 at 15 12 09" src="https://user-images.githubusercontent.com/88528265/197338437-cc65b32c-62d2-4341-a210-f698b8178ad6.png">

But we see different results for these cases
```bash
export REGIONS="eu-west-1 eu-west-2"

# strings.Split
gomplate -i '{{ strings.Split .Env.REGIONS " " }}'
[ ]

# strings.Split - pipeline
gomplate -i '{{ .Env.REGIONS | strings.Split " " }}'
[eu-west-1 eu-west-2]

# split
gomplate -i '{{ split .Env.REGIONS " " }}'
[eu-west-1 eu-west-2]

# split - pipeline
gomplate -i '{{ .Env.REGIONS | split " " }}'
[ ]
```

1. Is it correct that `split` can't be used with pipeline and `strings.Split` just with pipeline?
2. In provided example we got same result with or without parentheses - are they required or only for complex cases?
    ```bash
    gomplate -i '{{ range .Env.REGIONS | strings.Split " " }} {{ . }} {{ end }}'
    gomplate -i '{{ range (.Env.REGIONS | strings.Split " ") }} {{ . }} {{ end }}'
    ```

<details>
  <summary>Example</summary>
</br>

**Variable**
```bash
export REGIONS="eu-west-1 eu-west-2"
```

**Template**
```go
{{ range (split .Env.REGIONS " ") }}
provider "aws_split" {
  region = "{{.}}"
  alias  = "{{.}}"
}
{{ end }}

{{ range (.Env.REGIONS | split " ") }}
provider "aws_split_pipe" {
  region = "{{.}}"
  alias  = "{{.}}"
}
{{ end }}

{{ range (strings.Split .Env.REGIONS " ") }}
provider "aws_strings.Split" {
  region = "{{.}}"
  alias  = "{{.}}"
}
{{ end }}

{{ range (.Env.REGIONS | strings.Split " ") }}
provider "aws_strings.Split_pipe" {
  region = "{{.}}"
  alias  = "{{.}}"
}
{{ end }}
```

**Result**
```terraform
provider "aws_split" {
  region = "eu-west-1"
  alias  = "eu-west-1"
}

provider "aws_split" {
  region = "eu-west-2"
  alias  = "eu-west-2"
}



provider "aws_split_pipe" {
  region = " "
  alias  = " "
}



provider "aws_strings.Split" {
  region = " "
  alias  = " "
}



provider "aws_strings.Split_pipe" {
  region = "eu-west-1"
  alias  = "eu-west-1"
}

provider "aws_strings.Split_pipe" {
  region = "eu-west-2"
  alias  = "eu-west-2"
}
```
</details>

**Comment by hairyhenderson** [comment, marked as answer]
Hi @air3ijai, thanks for taking the time to post this.

> [...] And it looks like `strings.Split` works only with a pipe-line, [...]

No, it works both ways. The pipeline syntax is just syntactic sugar for providing the left-hand side of the `|` as the right-most argument.

I see in your example you've flipped the order of the two arguments, however:

```
# strings.Split
gomplate -i '{{ strings.Split .Env.REGIONS " " }}'
[ ]
```

What you actually want is `{{ strings.Split " " .Env.REGIONS }}`.

As for the difference between `strings.Split` and `split`, this is unfortunately by design. If you look at the documentation for each, you'll see [`split`](https://docs.gomplate.ca/functions/strings/#split) is not an alias for [`strings.Split`](https://docs.gomplate.ca/functions/strings/#strings-split), and `split` does redirect you to using `strings.Split` for a compatible version.

I don't know why `split` isn't deprecated - it should be, because in all cases `strings.Split` should be preferred as it works with pipelines.

I'm going to improve the documentation to make this difference more clear, and also deprecate the `split` function prior to later removal (and perhaps eventual re-introduction as an alias to `strings.Split`).

**Reply by hairyhenderson** [reply]
I've updated the documentation in https://github.com/hairyhenderson/gomplate/pull/1538, and you can see a preview of the new documentation here: https://deploy-preview-1538--gomplate-docs.netlify.app/functions/strings/#strings-split

I'm going to merge it soon, but please let me know if you can think of anything else that would clarify - I can always change it again!

**Comment by ghost** [comment]
Hi @hairyhenderson, thank you for the clarification and documentation update!

Yes, looks like I missed the order and it works fine for now:

```bash
export REGIONS="eu-west-1 eu-west-2"

# strings.Split
gomplate -i '{{ strings.Split " " .Env.REGIONS }}'
[eu-west-1 eu-west-2]

# strings.Split - pipeline
gomplate -i '{{ .Env.REGIONS | strings.Split " " }}'
[eu-west-1 eu-west-2]
```

**Reply by hairyhenderson** [reply]
Awesome, glad to hear it's working for you!
